File: singleton-design-pattern/JAVA/PaymentGateway.java
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

class PaymentGatewayManager {
    private PaymentGatewayManager() {
        System.out.println("Payment Gateway Manager initialized!");
    }

    private static PaymentGatewayManager instance;

    private static Lock mtx = new ReentrantLock();

    public static PaymentGatewayManager getInstance() {
        if(instance == null) { // First check without locking
            mtx.lock(); // Acquire the lock before creating the instance
            try {
                if(instance == null) { // Double-checked locking
                    instance = new PaymentGatewayManager();
                }
            } finally {
                mtx.unlock(); // Always release the lock
            }
        }

        return instance;
    }

    public void processPayment(double amount) {
        System.out.println("Processing payment of $" + amount + " through the payment gateway.");
    }
}

class PaymentGateway {
    public static void main(String[] args) {
        PaymentGatewayManager paymentGatewayManager = PaymentGatewayManager.getInstance();

        paymentGatewayManager.processPayment(100.0);

        PaymentGatewayManager anotherPaymentGatewayManager = PaymentGatewayManager.getInstance();

        if(paymentGatewayManager == anotherPaymentGatewayManager){
            System.out.println("Singleton is working fine! Both the instances are same.");
        }
        else{
            System.out.println("Singleton is not working fine! Both the instances are different.");
        }
    }
}

File: singleton-design-pattern/Typescript/payment-gateway.ts
class MutexLock {
    private isLocked = false;
  
    acquire() {
      while (this.isLocked) {
        // Busy wait (not ideal for long-running locks)
      }
      this.isLocked = true;
    }
  
    release() {
      this.isLocked = false;
    }
} 

class PaymentGatewayManager {
    
    constructor() {
        console.log("Payment Gateway Manager initialized!")
    }
    
    private static instance: PaymentGatewayManager;
    // The lock object to ensure thread safety
    private static lock = new Object();
    
    public static getInstance() {
        if (!PaymentGatewayManager.instance) {
            // Locking to ensure that only one instance is created
            synchronized(PaymentGatewayManager.lock, () => {
                // Double-check to ensure that instance is not created by another thread
                if (!PaymentGatewayManager.instance) {
                    PaymentGatewayManager.instance = new PaymentGatewayManager();
                }
            });
        }
        return PaymentGatewayManager.instance;
    }

    public processPayment(amount: number){
        console.log(`Processing payment of $${amount} through the payment gateway.`)
    }
}

// Helper function to simulate synchronized blocks
function synchronized(lock: Object, callback: () => void) {
    // Simulate synchronization using a mutex-like pattern
    // In a real TypeScript/JavaScript environment, this would require async/await or a library
    callback();
}

class PaymentGateway {
    static main() {
        const paymentGatewayManagerInstance: PaymentGatewayManager = PaymentGatewayManager.getInstance();
        
        paymentGatewayManagerInstance.processPayment(100.0);
        
        const anotherPaymentGatewayManagerInstance: PaymentGatewayManager = PaymentGatewayManager.getInstance();

        if(paymentGatewayManagerInstance === anotherPaymentGatewayManagerInstance) {
            console.log("Singleton is working fine! Both the instances are same.");
        }
        else {
            console.log("Singleton is not working fine! Both the instances are different.");
        }
    }
}

PaymentGateway.main();